### PR TITLE
feat: Recent flags

### PR DESCRIPTION
**Pull request by kwasniew**
## About the changes

Shows recently used flags
Screenshot 2024-05-29 at 14 32 56

Details:
when you visit a feature view you store it in the local storage. Up to 3 flags are stores. The change is propagated using custom event so that the sidebar can update automatically.

### CODE REVIEW

File: frontend/src/component/feature/FeatureView/FeatureOverview/FeatureOverview.tsx [hunk 1]
```diff
@@ -37,6 +41,10 @@ const FeatureOverview = () => {
         useHiddenEnvironments();
     const onSidebarClose = () => navigate(featurePath);
     usePageTitle(featureId);
+    const { setLastViewed } = useLastViewedFlags();
```
Comment (kwasniew) [inline frontend/src/component/feature/FeatureView/FeatureOverview/FeatureOverview.tsx:44]: every time you visit a feature update last viewed

File: frontend/src/hooks/useCustomEvent.ts [hunk 2]
```diff
@@ -8,10 +8,10 @@ export const useCustomEvent = (
     eventName: string,
     handler: (event: CustomEvent) => void,
 ) => {
-    const emitEvent = () => {
+    const emitEvent = useCallback(() => {
```
Comment (kwasniew) [inline frontend/src/hooks/useCustomEvent.ts:11]: give stable identity to this function so when it's used in useEffect deps it doesn't change

File: frontend/src/hooks/useLastViewedFlags.ts [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+
+    const setCappedLastViewed = useCallback(
+        (flag: { featureId: string; projectId: string }) => {
+            if (!flag.featureId || !flag.projectId) return;
+            if (lastViewed.find((item) => item.featureId === flag.featureId))
+                return;
+            const updatedLastViewed = removeIncorrect([...lastViewed, flag]);
+            setLastViewed(
+                updatedLastViewed.length > MAX_ITEMS
+                    ? updatedLastViewed.slice(-MAX_ITEMS)
```
Comment (kwasniew) [inline frontend/src/hooks/useLastViewedFlags.ts:47]: we could do more elaborate LRU cache but I prefer a simpler solution

File: frontend/src/hooks/useLastViewedFlags.ts [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+const removeIncorrect = (flags?: any[]): Flag[] => {
+    if (!Array.isArray(flags)) return [];
+    return flags.filter((flag) => flag.featureId && flag.projectId);
+};
+
+const localStorageItems = (key: string) => {
+    return removeIncorrect(getLocalStorageItem(key) || []);
+};
+
+export const useLastViewedFlags = () => {
```
Comment (kwasniew) [inline frontend/src/hooks/useLastViewedFlags.ts:19]: stores last 3 flags in local storage and propagates the changes to other usage of this hook

**Review by FredrikOseberg — APPROVED** [review]
LGTM

**Pull request merged**